### PR TITLE
Fixes for PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by karyna-tsymbal-atwix**
### Description (*)
Issues were found when trying to launch Magento, and running Unit Tests, on php-8.1. env.
This PR is fixing Zend code to be compatible with PHP 8.1.

### Related Issues
1. magento/magento2#34441

### CODE REVIEW

**Review by andrewbess — APPROVED** [review]

**Review by xmav — APPROVED** [review]

**Review by fascinosum — CHANGES_REQUESTED** [review]

File: library/Zend/Db/Adapter/Abstract.php [hunk 1]
```diff
@@ -928,7 +928,8 @@ public function quote($value, $type = null)
     public function quoteInto($text, $value, $type = null, $count = null)
     {
         if ($count === null) {
-            return str_replace('?', $this->quote($value, $type), $text);
+            $quote = $this->quote($value, $type) ?? '""';
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Adapter/Abstract.php:931]: hi @fascinosum , to be sure that there will be no broken Sql , e.g. we replace this
WHERE `name` != ? AND ...
with this
WHERE `name` != AND

File: library/Zend/Db/Adapter/Pdo/Abstract.php [hunk 2]
```diff
@@ -293,7 +293,7 @@ protected function _quote($value)
             return $value;
         }
         $this->_connect();
-        return $this->_connection->quote($value);
+        return $this->_connection->quote($value ?? '');
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Adapter/Pdo/Abstract.php:296]: @fascinosum connection is a \PDO object and PDO::quote requires string.

File: library/Zend/Oauth/Http/Utility.php [hunk 3]
```diff
@@ -210,8 +210,6 @@ public function generateTimestamp()
      */
     public static function urlEncode($value)
     {
-        $encoded = rawurlencode($value);
-        $encoded = str_replace('%7E', '~', $encoded);
-        return $encoded;
+        return $value ? str_replace('%7E', '~', rawurlencode($value)) : '';
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Oauth/Http/Utility.php]: thanks, fixed

File: library/Zend/Locale/Data.php [hunk 4]
```diff
@@ -982,7 +982,7 @@ public static function getContent($locale, $path, $value = false)
         if (is_array($value)) {
             $val = implode('_' , $value);
         }
-        $val = urlencode($val);
+        $val = $val ? urlencode($val) : '';
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Locale/Data.php]: thanks, fixed

File: library/Zend/Http/Response.php [hunk 5]
```diff
@@ -258,9 +258,10 @@ public function isRedirect()
     public function getBody()
     {
         $body = '';
+        $transferEncoding = $this->getHeader('transfer-encoding');
 
         // Decode the body if it was transfer-encoded
-        switch (strtolower($this->getHeader('transfer-encoding'))) {
+        switch ($transferEncoding ? strtolower($transferEncoding) : '') {
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Http/Response.php]: thanks, fixed

File: library/Zend/Db/Select.php [hunk 6]
```diff
@@ -817,6 +817,7 @@ protected function _join($type, $name, $cond, $cols, $schema = null)
             $tableName = $name;
             $correlationName = $this->_uniqueCorrelation($tableName);
         }
+        $tableName = $tableName ?? '';
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Select.php]: thanks, fixed as you suggested

File: library/Zend/Db/Adapter/Pdo/Abstract.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Adapter/Pdo/Abstract.php]: Thanks for the review, fixes in Zend_Db_Adapter_Abstract::quote as you suggested.

File: library/Zend/Db/Adapter/Abstract.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Adapter/Abstract.php]: Thanks for the review, fixes in Zend_Db_Adapter_Abstract::quote as you suggested.

File: library/Zend/Http/Client.php [hunk 7]
```diff
@@ -1041,7 +1041,7 @@ public function request($method = null)
                    if (! empty($query)) {
                        $query .= '&';
                    }
-                $query .= http_build_query($this->paramsGet, null, '&');
+                $query .= http_build_query($this->paramsGet);
```
Comment (fascinosum) [inline library/Zend/Http/Client.php:1044]: Should it be
```suggestion
                $query .= http_build_query($this->paramsGet, '', '&');
```
to guarantee the result

File: library/Zend/Validate/File/Extension.php [hunk 8]
```diff
@@ -121,9 +121,7 @@ public function setCase($case)
      */
     public function getExtension()
     {
-        $extension = explode(',', $this->_extension);
-
-        return $extension;
+        return ($this->_extension !== null) ? explode(',', $this->_extension) : [''];
```
Comment (fascinosum) [inline library/Zend/Validate/File/Extension.php:124]: Should we change the value for L137 instead?

**Review by fascinosum — CHANGES_REQUESTED** [review]

File: library/Zend/Db/Adapter/Abstract.php [hunk 9]
```diff
@@ -893,7 +894,7 @@ public function quote($value, $type = null)
                             (?:[eE][+-]?\d+)?    # optional exponent on decimals or octals
                           )
                         )/x',
-                        (string) $value, $matches)) {
+                        $value, $matches)) {
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Adapter/Abstract.php]: I restored typecasting , for the case of mixed value

File: library/Zend/Db/Statement.php [hunk 10]
```diff
@@ -134,7 +134,7 @@ protected function _prepare($sql)
      */
     protected function _parseParameters($sql)
     {
-        $sql = $this->_stripQuoted($sql);
+        $sql = ($sql !== null) ? $this->_stripQuoted($sql) : '';
 
         // split into text and params
         $this->_sqlSplit = preg_split('/(\?|\:[a-zA-Z0-9_]+)/',
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Statement.php]: fixed

File: library/Zend/Http/Client.php (hunk 7 above)
Comment (karyna-tsymbal-atwix) [reply library/Zend/Http/Client.php]: thanks, fixed as you suggested

File: library/Zend/Validate/File/Extension.php (hunk 8 above)
Comment (karyna-tsymbal-atwix) [reply library/Zend/Validate/File/Extension.php]: thanks, fixed as you suggested

File: library/Zend/Db/Statement.php [hunk 11]
```diff
@@ -134,14 +134,17 @@ protected function _prepare($sql)
      */
     protected function _parseParameters($sql)
     {
-        $sql = $this->_stripQuoted($sql);
```
Comment (fascinosum) [inline library/Zend/Db/Statement.php:137]: I recommend leaving it as it is

File: library/Zend/Db/Statement.php [hunk 12]
```diff
@@ -134,14 +134,17 @@ protected function _prepare($sql)
...
-
-        // split into text and params
-        $this->_sqlSplit = preg_split('/(\?|\:[a-zA-Z0-9_]+)/',
-            $sql, -1, PREG_SPLIT_DELIM_CAPTURE|PREG_SPLIT_NO_EMPTY);
+        if ($sql !== null) {
```
Comment (fascinosum) [inline library/Zend/Db/Statement.php:137]: How will the following `foreach` be executed if `$sql === null`?

File: library/Zend/Validate/File/Extension.php [hunk 13]
```diff
@@ -134,7 +132,6 @@ public function getExtension()
      */
     public function setExtension($extension)
     {
-        $this->_extension = null;
```
Comment (fascinosum) [inline library/Zend/Validate/File/Extension.php:137]: I think
```
$this->_extension = '';
```
is more appropriate

**Review by fascinosum — CHANGES_REQUESTED** [review]

File: library/Zend/Validate/File/Extension.php (hunk 13 above)
Comment (andrewbess) [reply library/Zend/Validate/File/Extension.php:137]: Hello @fascinosum 
The property `$_extension` is `''` by default

Comment (fascinosum) [reply library/Zend/Validate/File/Extension.php:137]: Only right after initialization of the object

File: library/Zend/Db/Statement.php (hunk 12 above)
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Statement.php:138]: fixed

File: library/Zend/Validate/File/Extension.php (hunk 13 above)
Comment (karyna-tsymbal-atwix) [reply library/Zend/Validate/File/Extension.php:137]: done

**Review by fascinosum — APPROVED** [review]

File: library/Zend/Db/Statement/Pdo.php [hunk 14]
```diff
@@ -245,7 +245,7 @@ public function _execute(array $params = null)
      * @return mixed Array, object, or scalar depending on fetch mode.
      * @throws Zend_Db_Statement_Exception
      */
-    public function fetch($style = null, $cursor = null, $offset = null)
+    public function fetch($style = null, $cursor = PDO::FETCH_ORI_NEXT, $offset = 0)
```
[the remark this answers is not in the document]
Comment (karyna-tsymbal-atwix) [reply library/Zend/Db/Statement/Pdo.php]: hi @fascinosum , done

**Pull request merged**